File: 0110_Balanced_Binary_Tree/solution.java
package cn.huang.leetcode;

import cn.huang.leetcode.common.TreeNode;

/*
    110. Balanced Binary Tree
    Given a binary tree, determine if it is height-balanced.

    For this problem, a height-balanced binary tree is defined as:
      a binary tree in which the depth of the two subtrees of every node never differ by more than 1.
 */
public class LC_0110_BalancedBinaryTree {
    public boolean isBalanced(TreeNode root) {
        if (root == null)
            return true;

        int left = depth(root.left);
        int right = depth(root.right);
        if (left - right > 1 || left - right < -1) {
            return false;
        } else {
            return isBalanced(root.left) && isBalanced(root.right);
        }
    }

    private int depth(TreeNode n) {
        if (n == null) {
            return 0;
        }
        if (n.left == null && n.right == null) {
            return 1;
        } else {
            int left = depth(n.left);
            int right = depth(n.right);
            return 1 + (left > right ? left : right);
        }
    }
}

File: 0110_Balanced_Binary_Tree/solution.cpp
/*
110. Balanced Binary Tree
https://leetcode.com/problems/balanced-binary-tree/
 */
#include <algorithm>
using namespace std;

#include "common/TreeNode.h"

/*
get height
Solution 1: O(nlogn)
 */
class Solution
{
public:
  bool isBalanced(TreeNode *root)
  {
    if (!root)
      return true;
    int left_height = height(root->left);
    int right_height = height(root->right);
    return (abs(left_height - right_height) <= 1) && isBalanced(root->left) && isBalanced(root->right);
  }

private:
  int height(TreeNode *root)
  {
    if (!root)
      return 0;
    return max(height(root->left), height(root->right)) + 1;
  }
};

// O(n)
class Solution
{
public:
  bool isBalanced(TreeNode *root)
  {
    if (!root)
      return true;
    bool balanced = true;
    height(root, &balanced);
    return balanced;
  }

private:
  int height(TreeNode *root, bool *balanced)
  {
    if (!root)
      return 0;
    int left_height = height(root->left, balanced);
    if (!balanced)
      return -1;
    int right_height = height(root->right, balanced);
    if (!balanced)
      return -1;
    if (abs(left_height - right_height) > 1)
    {
      *balanced = false;
      return -1;
    }
    return max(left_height, right_height) + 1;
  }
};